### PR TITLE
feat(ses): support RedirectStaticModuleInterface without specifying a record

### PR DESCRIPTION
**Pull request by kumavis**
allow `importHook` to return an alias record that `ses/src/module-load.js` can attempt to import
this allows `importHook` to resolve to something that may be provided by a different part of the import phase, like the `moduleMapHook`

### CODE REVIEW

File: packages/ses/index.d.ts [hunk 1]
```diff
@@ -69,8 +69,14 @@ export interface RedirectStaticModuleInterface {
   importMeta?: any;
 }
 
+export interface AliasModuleInterface {
+  alias: string;
+  compartment: Compartment;
```
Comment (naugtur) [inline packages/ses/index.d.ts:74]: why would the alias come from a different compartment if the goal for this is to support the `browser` field? 
is this the part where browser field defines a different package name for an import? I was thinking of the browser field pointing to a different export of a package. 
Why would an importHook know of a different compartment than its own?

TL;DR: what's the point of having `compartment` here?

File: packages/ses/index.d.ts [hunk 2]
```diff
@@ -69,8 +69,14 @@ export interface RedirectStaticModuleInterface {
...
+}
+
 export type StaticModuleType =
   | RedirectStaticModuleInterface
+  | AliasModuleInterface
```
Comment (naugtur) [inline packages/ses/index.d.ts:79]: would it be possible to use the Redirect type with slight modifications? Alias and Redirect seem conceptually close.
That would require a way to tell if it should go through `loadRecord` or `memoizedLoadWithErrorAnnotation` and potentially some simplification of `Redirect...`?   
I want to make sure that we're intentionally making sure the same module instance gets cached under both specifiers. (I'm not sure this is what other engines or bundlers are doing when things get aliased, they probably are)
A Redirect that says "look me up again in moduleMap under this other name before moving forward" is potentially a viable alternative. 

I'm not sure if that's a good idea, just trying to go through the consequences and play the devil's advocate or checking if they're intentional.

Comment (kriskowal) [reply packages/ses/index.d.ts:79]: I would also like to entertain the possibility that this could have the same shape as a redirect. The implementation is likely the same, but with separate handling for the cases where a redirect appears in the moduleMap/moduleMapHook or the importHook.

That is to say, if we rename `alias` to `specifier`, does this break? Does this require us to distinguish an alias/redirect record differently?

File: packages/ses/index.d.ts (hunk 1 above)
Comment (kriskowal) [reply packages/ses/index.d.ts:74]: This feature should be sufficiently general to also handle `imports` in `package.json`.

**Review by kriskowal — CHANGES_REQUESTED** [review]
Requesting further research into aligning the shapes of module descriptors to make usage easier to learn and remember if not standardize.

File: packages/ses/src/module-load.js [hunk 3]
```diff
@@ -175,27 +175,64 @@ const loadWithoutErrorAnnotation = async (
...
+      mapSet(moduleRecords, moduleSpecifier, aliasRecord);
+      return aliasRecord;
+    }
+
+    // check if this redirect with an explicit compartment
+    if (staticModuleRecord.compartment !== undefined) {
+      // ensure expected record shape
+      if (staticModuleRecord.importMeta !== undefined) {
+        throw new TypeError(
+          'Cannot redirect to an inexplicit record with a specified importMeta',
```
Comment (kriskowal) [inline packages/ses/src/module-load.js:215]:
```suggestion
          'Cannot redirect to an implicit record with a specified importMeta',
```

File: packages/ses/src/module-load.js [hunk 4]
```diff
@@ -175,27 +175,64 @@ const loadWithoutErrorAnnotation = async (
...
+        aliasCompartment,
+        aliasSpecifier,
+        aliasModuleRecord,
+        pendingJobs,
+        moduleLoads,
+        errors,
+        importMeta,
+      );
+      mapSet(moduleRecords, moduleSpecifier, aliasRecord);
+      return aliasRecord;
```
Comment (kriskowal) [inline packages/ses/src/module-load.js:207]: This is fine. But, reviewing this has given me a set of fresh eyes and there’s an existing defect. Having the memo set after await makes suspicious that there might be a race condition here. 

But, `loadRecord`, on closer inspection, is synchronous! We should get rid of `await` in `await loadRecord` and remove `async` from the `loadRecord` declaration.

There remains a possibility here that concurrent calls to `importHook` that redirect to the same `specifier` with a `record` will race to define that module (the last one winning), and then an `importHook` for the aliased specifier might load it again. Fixing _that_ problem is definitely out of scope for this change. The proper solution to something like that is probably to thread a cancellation token thru `importHook` and ensure the _first_ attempt to define the aliased module wins, aborting and ignoring any that remain. (Way out of scope. I’ll make an issue.)

**Review by kriskowal — APPROVED** [review]
Thank you, this is good.
